open-apparel-registry/open-apparel-registry
# Enable feature flags in Django & React using django-waffle

## Overview

- add & configure django-waffle library
- create api-feature-flags API endpoint
- add feature flags Redux actions & reducers to store currently active feature flags in state
- create FeatureFlag component to accept a flag and check whether it is active in the feature flags state, rendering only if so
- add temporary "Claim a facility" text on FacilityDetailsSidebar component to activate only when the claim_a_facility feature is active
- set some default props in the NavbarDropdown component to fix two prop types warnings

Connects #514

## Demo

With `claim_a_facility` switch active:

![Screen Shot 2019-05-21 at 9 47 25 PM](https://user-images.githubusercontent.com/4165523/58141637-0a905f00-7c12-11e9-8953-4103e535c5c0.png)

->

![Screen Shot 2019-05-21 at 9 44 16 PM](https://user-images.githubusercontent.com/4165523/58141589-e896dc80-7c11-11e9-908c-770b93dee3ba.png)

With `claim_a_facility` switch inactive:

![Screen Shot 2019-05-21 at 9 44 47 PM](https://user-images.githubusercontent.com/4165523/58141592-ec2a6380-7c11-11e9-9bc6-c5d7485c46f0.png)

->

![Screen Shot 2019-05-21 at 9 45 04 PM](https://user-images.githubusercontent.com/4165523/58141598-ef255400-7c11-11e9-8988-92baa40e40f9.png)

With no `claim_a_facility` switch:

![Screen Shot 2019-05-21 at 9 45 21 PM](https://user-images.githubusercontent.com/4165523/58141605-f2204480-7c11-11e9-98f5-5ffb4679a27f.png)

->

![Screen Shot 2019-05-21 at 9 45 32 PM](https://user-images.githubusercontent.com/4165523/58141610-f51b3500-7c11-11e9-9c2c-0c50038a13f7.png)

## Notes

We discussed potentially using this library https://github.com/garbles/flag but after investigating a bit further it seems like the functionality we needed from it was fairly straightforward to implement on our own. Since the library's still pre-official-release it seemed preferable to use our own implementation.

## Testing Instructions

- get this branch then `./scripts/update`
- `./scripts/manage resetdb`, `./scripts/manage processfixtures` and `./scripts/manage createsuperuser`
- sign in to the Django admin and verify that you now see entries for `Django-Waffle`
- without setting a switch, visit the React app and then look at the details page for a facility. Verify that you do not see the "Claim this facility." text at the bottom of the sidebar
- back in the Django admin, add a `claim_a_facility` switch but do not yet activate it, then save the new object
- return to the React facility details page and refresh, verifying that you still don't see the "Claim this facility." text
- back again in the Django admin, activate the `claim_a_facility` switch and then save the update.
- return again to the React facility details page and refresh, verifying that when the page fully loads you do see the "Claim this facility." text

## Checklist
